scalable-go-service/index.md

```diff
@@ -0,0 +1,322 @@
+	}
+
+	appRouter := mux.NewRouter()
+	appRouter.HandleFunc("/rates", RatesHandler(ratesRQChan))
+	srv := http.Server{
+		Addr:    viper.GetString("server.host") + ":" + viper.GetString("server.port"),
+		Handler: appRouter,
+	}
+
+	wg.Add(1)
+	go func() {
+		tokenLoader.Start()
+		wg.Done()
+	}()
+
+	wg.Add(1)
+	go func() {
+		rateLoaderPool.Start()
+		wg.Done()
+	}()
+
+    // omitting implementation of gracefully shutting down all the workers
+}
+```
+
+#### Rates Route Handler
+
+```go
+func RatesHandler(ratesRQChan chan domain.RateRQ) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+	    var rates []types.HotelRateInfo
+	    rateRSChan := make(chan RateRS, len(r.HotelIds))
+	    for _, hotelID := range r.HotelIds {
+	    	ratesRQChan <- RateRQ{
+	    		HotelID:                 hotelID,
+	    		CheckInDate:             r.CheckInDate,
+	    		CheckOutDate:            r.CheckOutDate,
+	    		RateRSChan:              rateRSChan,
+	    	}
+	    }
+	    for range p.HotelIds {
+	    	res := <-rateRSChan
+            rates = append(rates, res.HotelRateInfo)
+	    }
+		writeResponse(&w, createJsonResponse(r.Context(), rates, http.StatusOK))
+	}
+}
+```
+
+#### Rate Loader Pool
+
+```go
+type RateLoaderPool struct {
+	Count         int
+	In            chan RateRQ
+	TokenRQChan   chan<- TokenRQ
+	stop          chan struct{}
+}
+
+// Start initializes and starts n unique rate loader based on the Count parameter.
+// Each rate loader listens for requests on the pool's In channel.
+// For each request, a loader asks for a token through the TokenRQChan,
+// gets rates from the vendor, and puts the result on the requests'
+// desired response channel (RateRSChan).
+func (rlp *RateLoaderPool) Start() {
+	rlp.stop = make(chan struct{}, 1)
+	var wg sync.WaitGroup
+	for i := 0; i < rlp.Count; i++ {
+		wg.Add(1)
+		tokenRSChan := make(chan TokenRS)
+		go func() {
+		loop:
+			for {
+				select {
+				case rq := <-rlp.In:
+
+					// ask for a token from a token loader
+					rlp.TokenRQChan <- TokenRQ{...}
+					token := <-tokenRSChan
+
+					// get rates for a hotel
+					r := getRateFromVendor(...)
+
+					rq.RateRSChan <- RateRS{r}
+				case <-rlp.stop:
+					break loop
+				}
+			}
+			wg.Done()
+		}()
+	}
+    // blocks until all workers in the pool have stopped
+	wg.Wait()
+}
+
+// Stop stops all workers in the pool
+func (rlp *RateLoaderPool) Stop() {
+	close(rlp.stop)
+}
+```
+
+#### Token Loader
+
+```go
+type TokenLoader struct {
+	In   <-chan TokenRQ
+	stop chan struct{}
+}
+
+// Start initializes and starts the token loader
+func (tl *TokenLoader) Start() {
+	tl.stop = make(chan struct{}, 1)
+	var token *TokenRS
+loop:
+	for {
+		select {
+		case rq := <-tl.In:
+			rq.TokenRSChan <- getAuthenticationToken(...)
+		case <-tl.stop:
+			break loop
+		}
+	}
+}
+
+// Stop token loader
+func (tl *TokenLoader) Stop() {
+	close(tl.stop)
+}
+```
+
+## Importance of Communication
```

Comment:
This section is confusing. On a meta level it is funny, but I don't think that is the intention.